doc/HLD_of_FOL.rst

```diff
@@ -0,0 +1,290 @@
+
+Overview
+=========
+
+FOL is stored in a transactional container [1] populated with records indexed [2] by LSN. An LSN is used to refer to a point in FOL from other meta-data tables (epochs table, object index, sessions table, etc.). To make such references more flexible, a FOL, in addition to genuine records corresponding to updates, might contain pseudo-records marking points on interest in the FOL to which other file system tables might want to refer to (for example, an epoch boundary, a snapshot origin, a new server secret key, etc.). By abuse of terminology, such pseudo-records will be called FOL records too. Similarly, as part of redo-recovery implementation, DTM might populate a node FOL with records describing updates to be performed on other nodes.
+
+[1][R.BACK-END.TRANSACTIONAL] ST
+
+[2][R.BACK-END.INDEXING] ST
+
+Record Structure
+=================
+
+A FOL record, added via the main FOL interface, contains the following:
+
+- an operation opcode, identifying the type of file system operation;
+
+- LSN;
+
+- information sufficient to undo and redo the update, described by the record, including:
+
+  - for each file system object affected by the update, its identity (a fid) and its object version identifying the state of the object in which the update can be applied;
+
+  - any additional operation type dependent information (file names, attributes, etc.) necessary to execute or roll-back the update;
+
+- information sufficient to identify other updates of the same operation (if any) and their state. For the purposes of the present design specification it's enough to posit that this can be done by means of some opaque identifier;
+
+- for each object modified by the update, a reference (in the form of lsn) to the record of the previous update to this object (null is the update is object creation). This reference is called prev-lsn reference;
+
+- distributed transaction management data, including an epoch this update and operation are parts of;
+
+- liveness state: a number of outstanding references to this record
+
+Liveness and Pruning
+=====================
+
+A node FOL must be prunable if only to function correctly on a node without persistent storage. At the same time, a variety of sub-systems both from M0 core and outside of it, might want to refer to FOL records. To make pruning possible and flexible, each FOL record is augmented with a reference counter, counting all outstanding references to the record. A record can be pruned if its reference counter drops to 0 together with reference counters of all earlier (in lsn sense) unpruned records in the fol.
+
+Conformance
+=============
+
+- [R.FOL.EVERY-NODE]: on nodes with persistent storage, M0 core runs in the user space and the FOL is stored in a data-base table. On a node without persistent storage, or M0 core runs in the kernel space, the fol is stored in memory-only index. Data-base and memory-only index provide the same external interface, making FOL code portable;
```

Comment:
Please capitalize the abbreviation fol to FOL.

Reply:
Incorporated.